modules/fluid_properties/include/fluidproperties/IdealGasMixtureFluidProperties.h

```diff
@@ -0,0 +1,200 @@
+  Real mixtureMolarMass(const std::vector<Real> & x) const;
+
+  /**
+   * Computes molar fractions
+   *
+   * @param[in] x   All mass fractions
+   */
+  template <typename CppType>
+  std::vector<CppType> molarFractionsFromMassFractions_templ(const std::vector<CppType> & x) const;
+  std::vector<ADReal> molarFractionsFromMassFractions(const std::vector<ADReal> & x) const;
+  std::vector<Real> molarFractionsFromMassFractions(const std::vector<Real> & x) const;
+
+  override_property(p, v, e);
+  override_property(T, v, e);
+  override_property(v, p, T);
+  override_property(e, p, T);
+  override_property(s, p, T);
+  override_property(c, p, T);
+  override_property(cp, p, T);
+  override_property(cv, p, T);
+  override_property(mu, p, T);
+  override_property(k, p, T);
+
+#undef override_property
+
+protected:
+  /// Names of component fluid properties
+  const std::vector<UserObjectName> _component_fp_names;
+  /// Number of components
+  const unsigned int _n_components;
+  /// Number of secondary components
+  const unsigned int _n_secondary_components;
+
+  /// Component fluid properties objects
+  std::vector<const IdealGasFluidProperties *> _component_fps;
+};
+
+#pragma GCC diagnostic pop
+
+template <typename CppType>
+std::vector<CppType>
+IdealGasMixtureFluidProperties::secondaryToAllMassFractions_templ(
+    const std::vector<CppType> & x_secondary) const
+{
+  mooseAssert(x_secondary.size() == _n_secondary_components, "Size mismatch");
+
+  CppType sum = 0;
+  for (const auto i : make_range(_n_secondary_components))
+    sum += x_secondary[i];
+
+  const CppType x_primary = 1.0 - sum;
+
+  std::vector<CppType> x;
+  x.push_back(x_primary);
+  x.insert(x.end(), x_secondary.begin(), x_secondary.end());
+
+  return x;
+}
+
+template <typename CppType>
+CppType
+IdealGasMixtureFluidProperties::mixtureMolarMass_templ(const std::vector<CppType> & x) const
+{
+  mooseAssert(x.size() == _n_components, "Size mismatch");
+
+  CppType sum = 0;
+  for (const auto i : make_range(_n_components))
+    sum += x[i] / _component_fps[i]->molarMass();
+
+  return 1.0 / sum;
+}
+
+template <typename CppType>
+std::vector<CppType>
+IdealGasMixtureFluidProperties::molarFractionsFromMassFractions_templ(
+    const std::vector<CppType> & x) const
+{
+  mooseAssert(x.size() == _n_components, "Size mismatch");
+
+  const auto M = mixtureMolarMass(x);
+  std::vector<CppType> psi(_n_components);
+  for (const auto i : make_range(_n_components))
+    psi[i] = x[i] * M / _component_fps[i]->molarMass();
+
+  return psi;
+}
+
+template <typename CppType>
+CppType
+IdealGasMixtureFluidProperties::p_from_v_e_templ(const CppType & v,
+                                                 const CppType & e,
+                                                 const std::vector<CppType> & x_secondary) const
+{
+  const auto x = secondaryToAllMassFractions(x_secondary);
+  const auto M = mixtureMolarMass(x);
+
+  return _R * T_from_v_e(v, e, x_secondary) / (M * v);
```

Comment:
```suggestion
  return M * _R * T_from_v_e(v, e, x_secondary) / (v);
```
?
pv = MRT

Reply:
I think you're thinking of this:
p V = m Rspecific T
p v = Rspecific T
p v = (Runiversal / M) T